### PR TITLE
[IMP] fleet new api 

### PR DESCRIPTION
**Pull request by DudhatShivam**
Task: https://www.odoo.com/web#id=10075&view_type=form&model=project.task&menu_id=3942&action=327

Pad:https://pad.odoo.com/p/r.xFp6558VZvondGaA

### CODE REVIEW

File: addons/fleet/__openerp__.py [hunk 1]
```diff
@@ -25,22 +25,21 @@
...
-        'fleet_cars.xml',
-        'fleet_data.xml',
-        'fleet_board_view.xml',
+        'views/fleet_view.xml',
+        'data/fleet_cars_data.xml',
+        'data/fleet_data.xml',
     ],
 
-    'demo': ['fleet_demo.xml'],
+    'demo': ['data/fleet_demo_data.xml'],
```
Comment (rgo-odoo) [inline addons/fleet/__openerp__.py]: demo files should be post-fixed with _demo.xml(not with _data.xml)

File: addons/fleet/models/fleet_vehicle.py [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+    _order = 'license_plate asc'
+    _inherit = 'mail.thread'
+
+    @api.model
+    def default_stage(self):
+        try:
+            stage = self.env.ref('fleet.vehicle_stage_active')
+        except ValueError:
+            stage = False
+        return stage
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: remove try..expect and replace it with,
return self.env.ref('fleet.vehicle_stage_active', raise_if_not_found=False)

File: addons/fleet/models/fleet_make.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+
+from openerp import api, fields, models, tools
+
+
+class FleetMake(models.Model):
+    _name = 'fleet.make'
+    _description = 'Make of the vehicle'
+    _order = 'name asc, id asc'
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_make.py]: no need to define explicit ascending order, _order = 'name, id' is enough(apply this at all the places in this branch)

File: addons/fleet/models/fleet_make.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+                                      "resize as a 128x128px image, with aspect ratio preserved. "
+                                      "Use this field in form views or some kanban views.")
+    image_small = fields.Binary(compute='_compute_get_image', inverse='_compute_image', string="Small-sized photo", store=True,
+                                help="Small-sized photo of the brand. It is automatically "
+                                     "resize as a 64x64px image, with aspect ratio preserved. "
+                                     "Use this field anywhere a small image is required.")
+
+    @api.one
+    @api.depends('image')
+    def _compute_get_image(self):
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_make.py]: rename _compute_get_image(...) --> _compute_image(...)

File: addons/fleet/models/fleet_make.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        make_images = tools.image_get_resized_images(self.image)
+        self.image_medium = make_images.get('image_medium')
+        self.image_small = make_images.get('image_small')
+
+    @api.one
+    def _compute_image(self):
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_make.py]: rename _compute_image(...) --> _inverse_image_medium(...)

File: addons/fleet/models/fleet_make.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    _description = 'Make of the vehicle'
+    _order = 'name asc, id asc'
+
+    name = fields.Char(string='Make', required=True)
+    image = fields.Binary(string="Logo", help="This field holds the image used as logo for the brand,limited to 1024x1024px.")
+    image_medium = fields.Binary(compute='_compute_get_image', inverse='_compute_image', string="Medium-sized photo", store=True,
+                                 help="Medium-sized logo of the brand. It is automatically "
+                                      "resize as a 128x128px image, with aspect ratio preserved. "
+                                      "Use this field in form views or some kanban views.")
+    image_small = fields.Binary(compute='_compute_get_image', inverse='_compute_image', string="Small-sized photo", store=True,
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_make.py]: add a method "_inverse_image_small(...)" and pass it in 'inverse' attribute

File: addons/fleet/models/fleet_vehicle.py [hunk 7]
```diff
@@ -0,0 +1,553 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from collections import defaultdict
+import datetime
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: from datetime import timedelta is enough here(make sure to replace datetime.timedelta() call with timedelta(...) in this file)

File: addons/fleet/models/fleet_vehicle.py [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+    def _compute_get_contract_reminder(self):
+        LogContract = self.env['fleet.vehicle.log.contract']
+        current_date = fields.Datetime.from_string(fields.Date.context_today(self))
+        for vehicle in self:
+            overdue = False
+            due_soon = False
+            total = 0
+            name = ''
+            for contract in LogContract.search([('vehicle_id', '=', vehicle.id), ('state', 'in', ('open', 'toclose')), ('expiration_date', '!=', None)],
+                                                limit=1, order='expiration_date asc'):
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: you may avoid a call to search(...) by replacing it with,

if vehicle.log_contracts_ids:
    contract = vehicle.log_contracts_ids.filtered(lambda contract:
                         contract.state in ('open', 'toclose') \
                         and contract.expiration_date != None).sorted(lambda contract: \
                         contract.expiration_date)[0]

File: addons/fleet/models/fleet_vehicle.py [hunk 9]
```diff
@@ -0,0 +1,549 @@
...
+
+from openerp import api, fields, models, _
+from openerp.exceptions import UserError
+
+
+class FleetVehicle(models.Model):
+
+    _name = 'fleet.vehicle'
+    _description = 'Information on a vehicle'
+    _order = 'license_plate asc'
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: remove asc

File: addons/fleet/models/fleet_vehicle.py [hunk 10]
```diff
@@ -0,0 +1,549 @@
...
+
+    _sql_constraints = [('unique_chassis_number', 'unique(vin_sn)', 'Same Chassis Number is already exists.')]
+
+    @api.one
+    @api.depends('license_plate')
+    def _compute_vehicle_name(self):
+        self.name = ('%s' + ' / ' + '%s' + ' / ' + '%s') % (self.model_id.make_id.name, self.model_id.name, self.license_plate)
+
+    @api.multi
+    def _compute_get_contract_reminder(self):
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: Follow the name pattern at all places in this branch for,

compute functions = _compute_<field_name>
inverse functions = _inverse_<field_name>
default functions = _default_<field_name>
onchange function = _onchange_<field_name>

File: addons/fleet/models/fleet_vehicle.py [hunk 11]
```diff
@@ -0,0 +1,549 @@
...
+
+            vehicle.contract_renewal_overdue = overdue
+            vehicle.contract_renewal_due_soon = due_soon
+            vehicle.contract_renewal_total = (total - 1)  # we remove 1 from the real total for display purposes
+            vehicle.contract_renewal_name = name
+
+    @api.multi
+    def _compute_count_odometer(self):
+        Odometer = self.env['fleet.vehicle.odometer']
+        odometer_count = Odometer.read_group([('vehicle_id', '=', self.id)], fields=['vehicle_id'], groupby=['vehicle_id'])
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: It is not safe to use self.id if you decorate your method with @api.multi, would you replace your code with,

odometer_count = self.env['fleet.vehicle.odometer'].read_group([('vehicle_id', 'in', self.ids)], fields=['vehicle_id'], groupby=['vehicle_id'])
vehicle_dict = dict((record['vehicle_id'][0], record['vehicle_id_count']) for record in odometer_count)
for vehicle in self:
    vehicle.odometer_count = vehicle_dict.get(vehicle.id)

Do similar stuff for all your *_count fields wherever search_count is used.

File: addons/fleet/models/fleet_vehicle.py [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+    def _search_get_overdue_contract_reminder(self, operator, value):
+        if not (operator in ('=', '!=', '<>') and value in (True, False)):
+            raise UserError(_("Operation not supported"))
+        if (operator == '=' and value is True) or (operator in ('<>', '!=') and value is False):
+            search_operator = 'in'
+        else:
+            search_operator = 'not in'
+        today = fields.Date.context_today(self)
+        overdue_contracts = self.env['fleet.vehicle.log.contract'].search([('expiration_date', '<', today), ('state', 'in', ('open', 'toclose'))])
+        overdue_contract_vehicle = [contract.vehicle_id.id for contract in overdue_contracts]
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: rename overdue_contract_vehicle --> overdue_contract_vehicle_ids

File: addons/fleet/models/fleet_vehicle.py [hunk 13]
```diff
@@ -0,0 +1,549 @@
...
+    vendor_id = fields.Many2one('res.partner', string='Supplier', domain="[('supplier', '=', True)]")
+    notes = fields.Text()
+    cost_id = fields.Many2one('fleet.vehicle.cost', string='Cost', required=True, ondelete='cascade')
+    cost_amount = fields.Float(related='cost_id.amount', string='Amount', store=True)
+    date = fields.Date(default=fields.Date.context_today)
+    # we need to keep this field as a related with store=True because the graph view doesn't support (1) to address
+    # fields from inherited table and (2) fields that aren't stored in database
+
+    @api.onchange('vehicle_id')
+    def on_change_vehicle(self):
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: rename on_change_vehicle --> _onchange_vehicle() and similar changes for below onchange methods,

File: addons/fleet/models/fleet_vehicle.py [hunk 14]
```diff
@@ -0,0 +1,549 @@
...
+    date = fields.Date(default=fields.Date.context_today)
+    value = fields.Float('Odometer Value', group_operator="max")
+    vehicle_id = fields.Many2one('fleet.vehicle', string='Vehicle', required=True)
+    unit = fields.Selection(related='vehicle_id.odometer_unit', readonly=True)
+
+    @api.one
+    @api.depends('vehicle_id', 'date')
+    def _compute_vehicle_log_name(self):
+        name = self.vehicle_id and self.vehicle_id.name or ''
+        self.name = self.date and name + ' / ' + self.date or name
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: name = self.vehicle_id.name
self.name = name + ' / ' + self.date if self.date else name

File: addons/fleet/models/fleet_vehicle.py (hunk 13 above)
Comment (nhomar) [inline addons/fleet/models/fleet_vehicle.py]: >> rename on_change_vehicle --> _onchange_vehicle() and similar changes for below onchange methods,

Just to understand: Why do you ask for this change?

Isn't now a good practice left the onchange public? if you put the `_` before this methods will not be callable thorught webservices AFAICAU.

It is just a curious question :-)

File: addons/fleet/models/fleet_vehicle_model.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    _description = 'Model of a vehicle'
+    _order = 'name, id '
+
+    @api.multi
+    def name_get(self):
+        return self.mapped(lambda m: (m.id, '/'.join(filter(None, (m.make_id.name, m.name)))))
+
+    name = fields.Char(required=True)
+    make_id = fields.Many2one('fleet.make', string='Make', oldname='brand_id', required=True, help='Make of the vehicle')
+    vendors = fields.Many2many('res.partner', 'fleet_vehicle_model_vendors', 'model_id', 'partner_id')
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle_model.py]: rename to vendor_ids and set oldname parameter to 'vendors'

File: addons/fleet/models/fleet_vehicle.py (hunk 13 above)
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: The onchange methods are invoked on a pseudo-record that contains the values present in the form and it doesn't return record value now(it deals with value in cache), so one isn't going to get any benefit if we make it public.

File: addons/fleet/models/fleet_vehicle.py [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+    # we need to keep this field as a related with store=True because the graph view doesn't support (1) to address
+    # fields from inherited table and (2) fields that aren't stored in database
+
+    @api.onchange('vehicle_id')
+    def on_change_vehicle(self):
+        self.odometer_unit = self.vehicle_id.odometer_unit
+        self.purchaser_id = self.vehicle_id.driver_id
+
+    @api.onchange('liter', 'price_per_liter', 'amount')
+    def on_change_liter(self):
```
Comment (rgo-odoo) [inline addons/fleet/models/fleet_vehicle.py]: on_change_liter(), on_change_price_per_liter() and on_change_amount() are having same logic!

**Comment by rim-odoo** [comment]
Replaced by https://github.com/odoo/odoo/pull/9058

**Pull request closed without merging**